src/lib/Imp.hs

```diff
@@ -181,6 +200,21 @@ toImpOp (maybeDest, op) = case op of
     extraOffset <- indexToInt (PairTy u v) (PairVal idx vz)
     tileOffset' <- iaddI (fromScalarAtom tileOffset) extraOffset
     returnVal $ toScalarAtom resultTy tileOffset'
+  ThrowError ty -> do
+    emitStatement (Nothing, IThrowError)
+    return $ Con $ AnyValue ty
+  CoerceOp destTy@PreludeBoolTy x -> case getType x of
+    BoolTy -> do
+      tag <- emitInstr $ IPrimOp $ ScalarUnOp BoolToInt $ fromScalarAtom x
+      returnVal $ Con $ SumAsProd destTy (toScalarAtom IntTy tag) [[], []]
+    ty -> error $ "unexpected type: " ++ pprint ty
+  CoerceOp BoolTy x -> case x of
+    DataCon _ [] 0 [] -> returnVal $ Con $ Lit $ BoolLit False
+    DataCon _ [] 1 [] -> returnVal $ Con $ Lit $ BoolLit True
+    Con (SumAsProd _ tag [[],[]]) -> do
```

Comment:
What's the difference between `DataCon` and `SumAsProd`? Do we use the first when we know the constructor statically, while the later is for the cases when it might be dynamic?

Reply:
Yes, that's right. But `SumAsProd` is a wart. We'll be able to get rid of it once we let ourselves use arbitrary (cheapish) expressions to reconstruct atoms from pointers, rather than having to pretend that everything is an atom.